src/python/pants/bin/remote_pants_runner.py

```diff
@@ -64,22 +99,17 @@ def __init__(self, exiter, args, env, options_bootstrapper, stdin=None, stdout=N
   def _trapped_signals(self, client):
     """A contextmanager that overrides the SIGINT (control-c) and SIGQUIT (control-\\) handlers
     and handles them remotely."""
-    def handle_control_c(signum, frame):
-      client.maybe_send_signal(signum, include_pgrp=True)
 
-    existing_sigint_handler = signal.signal(signal.SIGINT, handle_control_c)
-    # N.B. SIGQUIT will abruptly kill the pantsd-runner, which will shut down the other end
-    # of the Pailgun connection - so we send a gentler SIGINT here instead.
-    existing_sigquit_handler = signal.signal(signal.SIGQUIT, handle_control_c)
+    graceful_signal_handler = PailgunClientSignalHandler(client)
+    previous_signals = ExceptionSink.reset_graceful_signal_handler(graceful_signal_handler)
 
-    # Retry interrupted system calls.
-    signal.siginterrupt(signal.SIGINT, False)
-    signal.siginterrupt(signal.SIGQUIT, False)
     try:
       yield
     finally:
-      signal.signal(signal.SIGINT, existing_sigint_handler)
-      signal.signal(signal.SIGQUIT, existing_sigquit_handler)
+      for signum, handler in previous_signals.items():
```

Comment:
Having this logic externalized from `ExceptionSink` feels a little bit error prone.

Perhaps rather than returning a dict containing signal mappings, `reset_graceful_signal_handler` could return a tiny class with a `def restore` that would restore the previous signals? Or could it return an anonymous `GracefulSignalHandler`, resulting in a pattern like?:
```
previous_signal_handler = ExceptionSink.reset_graceful_signal_handler(signal_handler)
try:
  ..
finally:
  ExceptionSink.reset_graceful_signal_handler(previous_signal_handler)
```

Reply:
Done! Was also able to make `_trapped_signals()` in `remote_pants_runner.py` much simpler!